vllm/compilation/backends.py

```diff
@@ -164,106 +166,285 @@ def wrap_inductor(graph, example_inputs, additional_inductor_config):
+    use_cudagraph: bool  # the size is in capture_sizes
+
+    compiled: bool = False
+    runnable: Callable = None  # type: ignore
+    num_finished_warmup: int = 0
+    cudagraph: Optional[torch.cuda.CUDAGraph] = None
+    output: Optional[Any] = None
+
+
+class PiecewiseBackend:
+
+    def __init__(self,
+                 graph: fx.GraphModule,
+                 compilation_configs: CompilationConfig,
+                 graph_pool: Any,
+                 is_first_graph: bool = False):
+        """
+        The backend for piecewise compilation.
+        It mainly handles the compilation and cudagraph capturing.
+
+        We will compile `self.graph` once for the general shape,
+        and then compile for different shapes specified in
+        `compilation_configs.compile_sizes`.
+
+        Independently, we will capture cudagraph for different shapes.
+
+        If a shape needs both compilation and cudagraph, we will
+        compile it first, and then capture cudagraph.
+        """
+        self.graph = graph
+        self.compilation_configs = compilation_configs
+        self.graph_pool = graph_pool
+        self.is_first_graph = is_first_graph
+
+        self.compile_sizes: Set[int] = set(
+            self.compilation_configs.compile_sizes)
+        self.capture_sizes: Set[int] = set(
+            self.compilation_configs.capture_sizes)
+
+        self.compile_finished = False
+        self.first_run_finished = False
+
+        self.compiled_graph_for_general_shape: Callable = None  # type: ignore
+
+        self.sym_shape_indices: List[int] = []
+
+        # the entries for different shapes that we need to either
+        # compile or capture cudagraph
+        self.concrete_size_entries: Dict[int, ConcreteSizeEntry] = {}
+        for shape in self.compile_sizes.union(self.capture_sizes):
+            self.concrete_size_entries[shape] = ConcreteSizeEntry(
+                runtime_shape=shape,
+                need_to_compile=shape in self.compile_sizes,
+                use_cudagraph=shape in self.capture_sizes,
+            )
+
+    def __call__(self, *args) -> Any:
+
+        if not self.compile_finished:
+            self.compile_finished = True
+
+            # this is the first compilation, we will compile a graph with
+            # dynamic shape, as the caller will mark first dimension as dynamic
+
+            self.sym_shape_indices = [
+                i for i, x in enumerate(args) if isinstance(x, torch.SymInt)
+            ]
+
+            if self.compilation_configs.use_inductor:
+                # args here are example inputs (fake tensors)
+                if self.is_first_graph:
+                    logger.info("Compiling a graph for general shape")
+                self.compiled_graph_for_general_shape = wrap_inductor(
+                    self.graph, args,
+                    self.compilation_configs.inductor_compile_config)
+            else:
+                self.compiled_graph_for_general_shape = self.graph
```

Comment:
This is another place for `compile_for_shape`

Reply:
fixed in [a9ae1ae](https://github.com/vllm-project/vllm/pull/9715/commits/a9ae1aec2c812795bb4d1b76bb93ed4c0d1c8bc5)